rep-2016.rst

```diff
@@ -0,0 +1,735 @@
+The reference implementation for this derivation is in ``rosidl_generator_type_description``\ [#td_gen]_.
+
+Versioning the ``TypeDescription`` Message Type
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Given that the type description message interface has to be generic enough to support anything described in the ROS interfaces, there will be a need to add or remove fields over time in the type description message itself.
+This should be done in such a way that the fields are tick-tocked and deprecated properly.
+It is recommended to do this by having explicitly named versions of this interface, e.g. ``TypeDescriptionV2``, ``TypeDescriptionV3`` and so on.
+The initial version is named without a version identifier.
+Revisions may be inevitable, but there should be a default preference against modifying, to keep the number of these versions to a minimum.
+
+``TypeDescription`` Self-Describing Loop Problem
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This section covers an implementation detail rather than a goal-oriented design decision, but is worth covering as it represents a significant difficulty.
+
+The ``TypeDescription`` messages are defined as just another ROS interface type, using all the same tooling.
+This means that its data structures are not available yet, during build time, to describe itself and other interface packages.
+The final decision to support this is a checked-in mirror of the generated type structs, provided in ``rosidl_runtime_c``, for use by C (used by Python) and C++ code generation.
+This decision incurs overhead and possibility for human error, if the interfaces are ever updated, as the runtime structs must be also updated with new generated code.
+To support this, infrastructure will have been added to automatically perform the updates, and check that they are updated properly in PR reviews.
+
+A few other options were considered but ultimately all were deemed too convoluted to follow even for those familiar with the ``rosidl`` type pipeline.
+
+Type Hash
+---------
+
+Hash Requirements
+^^^^^^^^^^^^^^^^^
+
+The following requirements are set for type hashes.
+
+Hashes for local types be available to client libraries without computation, thus must be precalculated at build time.
+
+The hash must be able to be calculated at runtime from information received on the wire.
+This allows subscribers to validate the received TypeDescriptions against advertised hashes, and allows dynamic publishers to invent new types and advertise their hash programmatically.
+
+The hash must only be computed using fields that affect communication compatibility, so that trivial changes do not change the hash
+Thus the hash excludes one aspect of Type Descriptions: it omits field default values.
+This is because default values for fields are only used by the writer of a packet, the recipient always receives some value in the field and can read it, thus defaults cannot affect compatibility.
```

Comment:
so after all, the constants are not supported at this moment. i understand the that is not requirement to check message field compatibility, but user application might check the value using constants on subscription? from user application perspective, this could be categorized as mismatch?

publisher sends the data using `DEFAULT_VALUE` as 1, but on the subscription using the same `DEFAULT_VALUE` because message constants are updated?

Reply:
This point does make sense to me. To put a full example:

Publisher's `My.msg`
```
uint8 GOOD_STATE = 1
uint8 BAD_STATE = 2

uint8 current_state
```

Publisher

```
My msg;
msg.current_state = My::GOOD_STATE;
publisher.publish(msg);
```

Subscription's `My.msg`
```
uint8 GOOD_STATE = 0
uint8 BAD_STATE = 1
uint8 UNKNOWN_STATE = 2

uint8 current_state
```

Subscription code

```
void message_callback(My msg) {
  if (msg.current_state == My::BAD_STATE)  {
    ERROR("Bad state!");
  }
}
```


This will trigger the bad state on the subscription, even though the publisher intended a good state. However, this doesn't currently trigger an update. 
(tangent: I think almost all constants in practice are enums, so it's unfortunate that we don't have enum support in messages)

@wjwwood I could use your thoughts on this, I'm forgetting if we discussed constants in the original design. They are not represented in `IndividualTypeDescription` so it's not that they are specifically ignored from the hash, they're not available to that calculation at all.

Reply:
Ah - we did not come to a final decisions, it was left as a TODO https://github.com/ros2/rcl_interfaces/tree/rolling/type_description_interfaces#todo